shared/actions/signup/index.js

```diff
@@ -156,219 +83,344 @@ function isValidName (name: ?string): ?string {
+    }
+  } else {
+    return {
+      type: Constants.checkPassphrase,
+      payload: {passphrase: new HiddenString(passphrase1)}
     }
+  }
+}
+
+export function resetSignup (): ResetSignup {
+  return {type: Constants.resetSignup, payload: {}}
+}
+
+export function sawPaperKey (): SawPaperKey {
+  return {type: Constants.sawPaperKey, payload: undefined}
+}
+
+type NodeCB = (err: ?any, result: ?any) => void
+
+// Wrappers around service calls
+function checkInviteCodeWithService (invitationCode: string, callback: NodeCB) {
+  const params: signup_checkInvitationCode_rpc = {
+    method: 'signup.checkInvitationCode',
+    param: {invitationCode},
+    incomingCallMap: {},
+    // TODO figure out waiting handler
+    waitingHandler: () => {},
+    callback
+  }
+  engine.rpc(params)
+}
+
+function checkUsernameAvailable (username: string, callback: NodeCB) {
+  const params: signup_checkUsernameAvailable_rpc = {
+    method: 'signup.checkUsernameAvailable',
+    param: {username},
+    waitingHandler: () => {},
+    incomingCallMap: {},
+    callback
+  }
+
+  engine.rpc(params)
+}
+
+function checkDeviceNameFormat (name: string, callback: NodeCB) {
+  const params: device_checkDeviceNameFormat_rpc = {
+    method: 'device.checkDeviceNameFormat',
+    param: {name},
+    waitingHandler: () => {},
+    incomingCallMap: {},
+    callback
+  }
+  engine.rpc(params)
+}
 
-    const params: signup_checkUsernameAvailable_rpc = {
-      method: 'signup.checkUsernameAvailable',
-      param: {username},
-      waitingHandler: isWaiting => dispatch(waiting(isWaiting)),
-      incomingCallMap: {},
-      callback: err => {
-        if (err) {
-          console.error("username isn't available:", err)
-          dispatch({
-            type: Constants.checkUsernameEmail,
-            error: true,
-            payload: {emailError, usernameError: `Username error: ${err.message}`, email, username}
-          })
-          resolve()
-        } else {
-          // We need this check to make flow happy. This should never be null
-          if (username && email) {
-            dispatch({
-              type: Constants.checkUsernameEmail,
-              payload: {username, email}
-            })
-            dispatch(nextPhase())
-            resolve()
-          } else {
-            reject()
-          }
+// This one is a bit different, since we don't want to change how everything in engine works just yet.
+// It's still a thunked actions
+function signupWithService (email: string, username: string, inviteCode: string, passphrase: HiddenString, deviceName: string, skipMail: boolean) {
+  return (dispatch, getState) => {
+    const params: signup_signup_rpc = {
+      method: 'signup.signup',
+      waitingHandler: () => {},
+      param: {
+        email,
+        inviteCode,
+        username,
+        deviceName,
+        passphrase: passphrase.stringValue(),
+        storeSecret: false,
+        skipMail
+      },
+      incomingCallMap: {
```

Comment:
If we decide to go this approach we could have the engine simply dispatch these like below instead of having to manually set that up.